src/11_6_2021_RcppClock_benchmarking_Rcpp_code.Rmd

```diff
@@ -0,0 +1,148 @@
+---
+title: "Benchmarking Rcpp code with RcppClock"
+author: "Zach DeBruine"
+license: GPL (>= 2)
+tags: Rcpp benchmarking RcppClock
+summary: Rcpp benchmarking with a simple C++ tick-tock clock and seamless integration
+  with the R environment at runtime.
+---
+
+Rcpp is all about the need for speed. However, Rcpp code is not magically fast, but is the result of careful coding and profiling. RcppClock makes it easy to profile C++ code from R:
+
+* On the C++ side, RcppClock wraps `std::chrono::high_resolution_clock` to time chunks of code, and exports the results to an R object in the global environment.
+* On the R side, RcppClock provides convenient methods for summarizing results and plotting with `ggplot2`.
+
+By benchmarking Rcpp code with RcppClock, we avoid measuring latency associated with converting R types to C++ types (and back again), and we can profile at line-by-line resolution rather than entire functions.
+
+Note: RcppClock is not compatible with OpenMP parallelization, and does not accurately measure processes taking less than ~1 microsecond.

```

Comment:
Can you expand?  I think some Chrono implementations do, and this may be implementation dependent. Or is it because of your choice of POSIXct?

Reply:
The accuracy of `std::chrono::high_resolution_clock` is implementation-dependent, but it is around 1 microsecond.  See https://stackoverflow.com/questions/48211688/how-accurate-is-stdchrono for some discussion.

We are just measuring intervals, and for benchmarking we just care about relative performance, so as long as the clock is steady, we are golden.